services/user/Tokens/service/src/lib.rs

```diff
@@ -0,0 +1,288 @@
+        token.check_is_owner(get_sender());
+        token.set_settings(index, enabled);
+    }
+
+    #[action]
+    fn open(token_id: TID) {
+        Balance::add(get_sender(), token_id);
+    }
+
+    #[action]
+    #[allow(non_snake_case)]
+    fn getBalance(token_id: TID, user: AccountNumber) -> Balance {
+        Balance::get_or_new(user, token_id)
+    }
+
+    #[action]
+    #[allow(non_snake_case)]
+    fn getSharedBal(
+        creditor: AccountNumber,
+        debitor: AccountNumber,
+        token_id: TID,
+    ) -> SharedBalance {
+        SharedBalance::get_or_new(creditor, debitor, token_id)
+    }
+
+    #[action]
+    fn recall(token_id: TID, from: AccountNumber, amount: Quantity, memo: String) {
+        check(amount.value > 0, "must be greater than 0");
+        let sender = get_sender();
+        let mut token = Token::get_assert(token_id);
+        let token_settings = token.settings();
+
+        check(sender != from, "cannot recall from self, use burn instead");
+        check(!token_settings.is_unrecallable(), "token is not recallable");
+
+        token.check_is_owner(sender);
+        token.burn(amount, from);
+
+        Wrapper::emit()
+            .history()
+            .recall(token_id, amount, sender, from, memo);
+    }
+
+    #[action]
+    fn burn(token_id: TID, amount: Quantity, memo: String) {
+        let sender = get_sender();
+        let mut token = Token::get_assert(token_id);
+        let token_settings = token.settings();
+
+        check(!token_settings.is_unburnable(), "token is not burnable");
+
+        token.burn(amount, sender);
+
+        Wrapper::emit()
+            .history()
+            .burn(token_id, sender, amount, memo);
+    }
+
+    #[action]
+    fn mint(token_id: TID, amount: Quantity, memo: String) {
+        let mut token = Token::get_assert(token_id);
+        let sender = get_sender();
+        token.check_is_owner(sender);
+        token.mint(amount, sender);
+
+        Wrapper::emit().history().minted(token_id, amount, memo);
+    }
+
+    #[action]
+    fn credit(token_id: TID, debitor: AccountNumber, amount: Quantity, memo: String) {
+        let creditor = get_sender();
+        SharedBalance::get_or_new(creditor, debitor, token_id).credit(amount);
+
+        Wrapper::emit()
+            .history()
+            .credited(token_id, creditor, debitor, amount, memo);
+    }
+
+    #[action]
+    fn uncredit(token_id: TID, debitor: AccountNumber, amount: Quantity, memo: String) {
+        let creditor = get_sender();
+
+        SharedBalance::get_or_new(creditor, debitor, token_id).uncredit(amount);
+
+        Wrapper::emit()
+            .history()
+            .uncredited(token_id, creditor, debitor, amount, memo);
+    }
+
+    #[action]
+    fn debit(token_id: TID, creditor: AccountNumber, amount: Quantity, memo: String) {
+        SharedBalance::get_or_new(creditor, get_sender(), token_id).debit(amount, memo);
+    }
+
+    #[event(history)]
+    pub fn recall(
```

Comment:
`burn` and `recall` events should be past tense `burned` and `recalled` for consistency with the other events and with the general convention